### DISCUSSION TITLE
SPA OAuth2 without Popups or second Window. Same Window login required

### DISCUSSION THREAD

**Discussion by MuslemRahimi** [pocketbase/pocketbase#3801] [Q&A] [answered]
Hi everyone,

I am building a oauth2 login in sveltekit+pocketbase.

For now it works but it opens a second window which i dont want that.

I know it is possible to build a workaround this issue see the Discussion here https://github.com/pocketbase/pocketbase/discussions/3718.

However, the author did not provide enough information for me to reproduce his result.

Any suggestions to fix the problem is much much appreciated
```
async function oauth2(provider:string) {
	let w = window.open()

	await pb
		.collection("users")
		.authWithOAuth2({
		provider: provider,
		urlCallback: (url) => {
			w.location.href = url
		},
		})
	toast.success('Login successfully!', {
		style: 'border-radius: 200px; background: #333; color: #fff;'});
	goto('/')
	
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
> ... it opens a second window which i dont want that.

But why? From users prespective as it is right now it should provide better UX because the operating context will be preserved and your app code can continue to work in a more "imperative" style.

In any case it is not possible at the moment to use the "all-in-one" OAuth2 flow and have the OAuth2 window open in the same page, because **the "all-in-one" flow relies that there will be a long lived connection between the browser and the PocketBase server during the OAuth2 authentication.**
Replacing the current page with another will prevent this from working, aka. your `goto('/')` call will never be executed because your SPA page will no longer be open.

If you really need to open the OAuth2 page in the same window, then you'll have to implement the "Manual code exchange" flow (see the second tab for plain JS example in https://pocketbase.io/docs/authentication/#oauth2-integration).
You'll need 2 frontend pages - one that will show the "Login with ..." links and another that will represent the redirect handler and will take care for the actual PocketBase authentication by calling `authWithOAuth2Code(...)` (_note that you'll have to update your OAuth2 app configuration to use the redirect page as redirect/callback url_).

**Reply by MuslemRahimi** [reply]
Thanks for the info, i need to understand this problem better.

The Why Question...

I use my website as a flutter webview app to ship as an Android/IOS App. The second window breaks something in the webview therefore i try to make it in the same window.

**Reply by ganigeorgiev** [reply]
Even if you open the OAuth2 page in the same WebView, note that some providers (eg. Google) discourages webviews and may even refuse to open their signin pages in a WebView (see also https://support.google.com/faqs/answer/12284343).

If you use Flutter you can try replacing your WebView with [`flutter_custom_tabs`](https://pub.dev/packages/flutter_custom_tabs) or other similar package that will take care for the cross platform handling.

**Reply by ganigeorgiev** [reply]
Also from https://developers.googleblog.com/2021/06/upcoming-security-changes-to-googles-oauth-2.0-authorization-endpoint.html:

> As part of that work, we recently introduced a new [secure browser policy](https://developers.google.com/identity/protocols/oauth2/policies#browsers) prohibiting Google OAuth requests in embedded browser libraries commonly referred to as embedded webviews. **All embedded webviews will be blocked starting on September 30, 2021.**

**Reply by MuslemRahimi** [reply]
Thank you so much for the links. I wondered why my App blocked the oauth2 of Google but not for Discord.

Now i know :)
